uspace/lib/c/generic/adt/array.c

```diff
@@ -0,0 +1,192 @@
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @addtogroup libc
+ * @{
+ */
+/** @file
+ *
+ * Implementation of dynamic array that grows or shrinks based upon no. of
+ * items it contains. Non-negligible part of implementation is in @ref
+ * array.h because of type genericity.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <adt/array.h>
+#include <macros.h>
+#include <mem.h>
+#include <stdlib.h>
+
+static errno_t array_realloc(array_t *da, size_t capacity)
+{
+	if (capacity == da->capacity) {
+		return EOK;
+	}
+
+	void *new_data = realloc(da->_data, da->_item_size * capacity);
+	if (new_data || capacity == 0) {
+		da->_data = new_data;
+		da->capacity = capacity;
+	}
+	return (new_data == NULL && capacity != 0) ? ENOMEM : EOK;
+}
+
+void array_destroy(array_t *da)
+{
+	array_clear(da);
+	free(da->_data);
+	da->capacity = 0;
+}
+
+/** Remove item at given position, shift rest of array */
+void array_remove(array_t *da, size_t index)
+{
+	assert(index < da->size);
+	_array_unshift(da, index, 1);
+	errno_t rc = array_reserve(da, da->size);
+	assert(rc == EOK);
+}
+
+/** Clear dynamic array (empty) */
+void array_clear(array_t *da)
+{
+	da->size = 0;
+}
+
+/** Clear subsequence of array
+ *
+ * @param[in/out]  da
+ * @param[in]      begin  index of first item to remove
+ * @param[in]      end    index behind last item to remove
+ */
+void array_clear_range(array_t *da, size_t begin, size_t end)
+{
+	assert(begin < da->size);
+	assert(end <= da->size);
+
+	_array_unshift(da, begin, end - begin);
+	errno_t rc = array_reserve(da, da->size);
+	assert(rc == EOK);
+}
+
+/** Concatenate two arrays
+ *
+ * @param[in/out]  da1  first array and concatenated output
+ * @param[in]      da2  second array, it is untouched
+ *
+ * @return EOK on success
+ * @return ENOMEM when allocation fails
+ */
+errno_t array_concat(array_t *da1, array_t *da2)
+{
+	assert(da1->_item_size == da2->_item_size);
+
+	errno_t rc = array_reserve(da1, da1->size + da2->size);
+	if (rc != EOK) {
+		return rc;
+	}
+	void *dst = da1->_data + da1->size * da1->_item_size;
+	void *src = da2->_data;
+	size_t size = da1->_item_size * da2->size;
+	memcpy(dst, src, size);
+	da1->size += da2->size;
+
+	return EOK;
+}
+
+/** Grows/shrinks array so that it effeciently stores desired capacity
+ *
+ * @param      da
+ * @param[in]  desired capacity of array (items)
+ *
+ * @return EOK
+ * @return ENOMEM
+ */
+errno_t array_reserve(array_t *da, size_t capacity)
+{
+	const size_t factor = 2;
```

Comment:
Agreed, the "algorithm" isnt really smart. My suggestion would be:

**bellow 100**
- increased by 10 units
- shrinked by 10 but only if there are is an capacity difference of 20 
- This means: the array has at least 10 units (no more NULL in data), shrinking is done more conservatively than increasing (prevents the problem which you addressed: "If it's the same, repeated addition and removal of a single element could cause reallocation on every step")

**above 100**
same as above - except we use the unit size of 100 instead of 10

**above 1000**
I would suggest that we stick to a max increase of 100 units. If we keep on increasing the unit size by some factor we end up having problem with large arrays as they will try to reallocate 10^16 at once while almost none of them will be used. A max buffer of 100 isnt particularly big, but it should be big enough

I would suggest to allow the array to be resized when shrinking but I see a problem with that. If you look at libraries from java/macosx then they allow you to initialize the array with an expected capacity. The way I suggest above would remove that expected capacity if you delete ONE element from the array. One way of solving this issue is to add extra functions which prevents this from happening but ... meh... I would prefer if the library "simply" works without having the coder to consider what's the best method. If you really have a fixed capacity it's probably better to use the old school c-arrays. Or we separate the library into 2 libraries. One which simply provides a struct for holding c-array as one type (no capacity) and another library which provides a capacity based array

Reply:
Alright, I pressed the "comment" button to quickly. I believe the way to go is indeed to create two libraries: array and array_capacity. The first one is a simple c-array wrapper à la:
```
typedef struct {
void *data;
size_t size;
} array_t;
```

Once initialized you have to resize it manually.

The other library wraps around this struct and does it automatically for you - but this can mean you lose some performance

```
typedef struct {
array_t *array;
size_t capacity;
} array_capacity_t;
```

this means that there will be duplicated functions like ```array_add()``` and ```array_capacity_add()```. Should be fine though as they actually do two different things

Any thoughts?

Reply:
That's not quite what my point was. Creating two libraries instead of one is definitely not something I'd like to do. What I meant is that you often want an auto-expanding array, but you rarely want an auto-shrinking one. It makes sense for high-level languages where memory management is always automatic and hidden, but in C, there are few cases where you want automatic shrinking like that. More often, it would be a reason to avoid the library altogether, which kind of defeats the purpose. If you want to conserve space, you really always want to shrink when the array is finished and to an exact size. In the cases where you remove and add elements during short-term processing, you absolutely don't want to shrink until you are finished, since that would just needlessly fragment memory -- remember that in C, fragmentation cannot really be fixed after the fact, in contrast to languages with moving garbage collectors. The only case I can think of where it would be helpful at all is dynamic stack, and for this case you'll want to use a purpose-made stack type anyway, instead of messing with a generic array.

Reply:
> More often, it would be a reason to avoid the library altogether, which kind of defeats the purpose.

That's something which crossed my mind too. Alternative I could remove the library and replace it with the static implementation as mention above and add two functions for converting between linked list and fixed array. ```array_convert_to_list()``` and vica versa


I just looked briefly into sysman and to why it uses a capacity array. I cannot find a good explanation - in fact, **list_t would be better**! There are no ```array_at()``` calls - indicating that having fast access is not needed.

Reply:
Could it be that the structures might be in multiple arrays of the same type at the same time? `list_t` is an intrusive structure, so it needs a link inside the element that can't be shared by multiple lists.

If that's not the case, converting sysman to use `list_t` seems like the better option. (^_^)

Reply:
> Could it be that the structures might be in multiple arrays of the same type at the same time?

You're right! There is a data structure which does reference itself https://github.com/matthieuriolo/helenos/blob/system_daemon/uspace/srv/sysman/job.h#L66-L73

But frankly, a simple "bucket" could replace this structure and we could remove the library all together

Reply:
Sounds good. If you think you can convert it without too much hassle, go for it! :)

Reply:
I think it should be possible :) sorry for the extra work though